File: packages/pvm-assemblyscript/assembly/host-functions/general/peek.ts
import { RESULT_CODE_PANIC } from '../../config'
import {
  ACCUMULATE_ERROR_OK,
  ACCUMULATE_ERROR_OOB,
  ACCUMULATE_ERROR_WHO,
  HostFunctionResult,
} from '../accumulate/base'
import {
  HostFunctionContext,
  HostFunctionParams,
  PeekPokeParams,
  PVMGuest,
  RefineInvocationContext,
} from './base'
import { BaseHostFunction } from './base'

/**
 * PEEK host function (Ω_P)
 *
 * Reads memory from a PVM machine instance
 *
 * *** GRAY PAPER FORMULA ***
 * Gray Paper: pvm_invocations.tex, Ω_P (peek = 9)
 *
 * Parameters: [n, o, s, z] = registers[7:4]
 * - n: machine ID
 * - o: destination offset in current memory
 * - s: source offset in machine's memory
 * - z: length
 *
 * Return states (equation 571-577):
 * - panic when Nrange{o}{z} not ⊆ writable[memory]  (destination not writable)
 * - continue with WHO when n not ∈ keys(m)  (machine doesn't exist)
 * - continue with OOB when Nrange{s}{z} not ⊆ readable{m[n].ram}  (source not readable)
 * - continue with OK otherwise
 *
 * Memory update: mem'[o:z] = (m[n].ram)[s:z]
 */
export class PeekHostFunction extends BaseHostFunction {
  functionId: u64 = u64(9) // PEEK function ID
  name: string = 'peek'
  gasCost: u64 = u64(10)

  execute(
    context: HostFunctionContext,
    params: HostFunctionParams | null,
  ): HostFunctionResult {
    if (!params) {
      return new HostFunctionResult(255) // continue execution
    }
    const peekParams = params as PeekPokeParams
    if (!peekParams.refineContext) {
      context.registers[7] = ACCUMULATE_ERROR_WHO
      return new HostFunctionResult(RESULT_CODE_PANIC)
    }

    // Gray Paper: [n, o, s, z] = registers[7:4]
    const machineId = u64(context.registers[7])
    const destOffset = u64(context.registers[8]) // o: destination
    const sourceOffset = u64(context.registers[9]) // s: source
    const length = u64(context.registers[10]) // z: length

    // Gray Paper error check order:
    // 1. Check if machine exists → WHO
    const refineContext = peekParams.refineContext!
    const machine = this.getPVMMachine(refineContext, machineId)
    if (!machine) {
      context.registers[7] = ACCUMULATE_ERROR_WHO
      return new HostFunctionResult(255) // continue (not HALT)
    }

    // 2. Check if source range is readable → OOB
    if (!this.isMachineMemoryReadable(machine, sourceOffset, length)) {
      context.registers[7] = ACCUMULATE_ERROR_OOB
      return new HostFunctionResult(255) // continue (not HALT)
    }

    // Read data from machine's memory
    const data = this.readFromMachineMemory(machine, sourceOffset, length)
    if (data === null) {
      context.registers[7] = ACCUMULATE_ERROR_OOB
      return new HostFunctionResult(255) // continue
    }

    // Gray Paper: mem'[o:z] = (m[n].ram)[s:z]
    const writeResult = context.ram.writeOctets(u32(destOffset), data)
    if (writeResult.hasFault) {
      return new HostFunctionResult(RESULT_CODE_PANIC)
    }

    // Return OK (0) for success
    context.registers[7] = ACCUMULATE_ERROR_OK
    return new HostFunctionResult(255) // continue execution
  }

  getPVMMachine(
    refineContext: RefineInvocationContext,
    machineId: u64,
  ): PVMGuest | null {
    // Get PVM machine by ID from context
    if (refineContext.machines.has(machineId)) {
      return refineContext.machines.get(machineId)!
    }
    return null
  }

  isMachineMemoryReadable(
    machine: PVMGuest,
    offset: u64,
    length: u64,
  ): bool {
    const result_readable = machine.pvm.state.ram.isReadableWithFault(
      u32(offset),
      u32(length),
    )
    const readable = result_readable.success
    const faultAddress = result_readable.faultAddress
    if (faultAddress !== 0) {
      return false
    }
    return readable
  }

  readFromMachineMemory(
    machine: PVMGuest,
    offset: u64,
    length: u64,
  ): Uint8Array | null {
    // Read data from machine's memory
    const readResult_data = machine.pvm.state.ram.readOctets(u32(offset), u32(length))
    const data = readResult_data.data
    const faultAddress = readResult_data.faultAddress
    if (data === null || faultAddress !== 0) {
      return null
    }
    return data
  }
}

File: packages/pvm-rust/src/host_functions/general/peek.rs
//! PEEK host function (Ω_P). Gray Paper: function ID 9.
//! r7=n (machine ID), r8=dest offset, r9=source offset in machine, r10=length.
//! No refineContext → set r7=WHO, return PANIC.

use crate::config::{FUNC_PEEK, REG_OK, REG_OOB, REG_WHO};
use crate::host_functions::base::{HostFunction, HostFunctionContext, HostFunctionResult};

/// PEEK (9): with refine_context: copy from machine RAM to current RAM; WHO if no machine, OOB if source unreadable; dest unwritable → PANIC.
pub struct PeekHostFunction;

impl HostFunction for PeekHostFunction {
    fn function_id(&self) -> u8 {
        FUNC_PEEK
    }
    fn name(&self) -> &'static str {
        "peek"
    }
    fn execute(&self, context: &mut HostFunctionContext<'_>) -> HostFunctionResult {
        let Some(refine) = context.refine_context.as_mut() else {
            context.registers[7] = REG_WHO;
            crate::host_log!("[hostfn] peek PANIC: no refine_context");
            return HostFunctionResult::panic();
        };

        let machine_id = context.registers[7];
        let dest_offset = context.registers[8] as u32;
        let source_offset = context.registers[9] as u32;
        let length = context.registers[10] as u32;

        let mut data = None;
        let found = refine.with_machine(machine_id, &mut |machine| {
            if machine.ram_is_readable(source_offset, length) {
                data = machine.ram_read(source_offset, length);
            }
        });
        if !found {
            context.registers[7] = REG_WHO;
            return HostFunctionResult::continue_execution();
        }
        let Some(data) = data else {
            context.registers[7] = REG_OOB;
            return HostFunctionResult::continue_execution();
        };

        let writable = context.ram.is_writable_with_fault(dest_offset, length);
        if !writable.success {
            crate::host_log!(
                "[hostfn] peek PANIC: dest not writable (offset={}, len={}, fault_address={})",
                dest_offset, length, writable.fault_address
            );
            return HostFunctionResult::panic();
        }
        let write_result = context.ram.write_octets(dest_offset, &data);
        if write_result.has_fault {
            crate::host_log!(
                "[hostfn] peek PANIC: dest write fault (offset={}, len={})",
                dest_offset, data.len()
            );
            return HostFunctionResult::panic();
        }

        context.registers[7] = REG_OK;
        HostFunctionResult::continue_execution()
    }
}
